Implement aggregated discovery API methods

doxxx93 opened the pull request:
Follow-up to #1873.

## Motivation

`Discovery::run()` currently requires N+2 API calls to discover all resources. With the Aggregated Discovery API added in #1873, we can now reduce this to just 2 calls.

## Solution

Added `Discovery::run_aggregated()` method that leverages the Aggregated Discovery API.

### Changes

- `Discovery::run_aggregated()` - New method using aggregated discovery
- `ApiGroup::from_v2()` - Converts `APIGroupDiscovery` to `ApiGroup`
- `GroupVersionData::from_v2()` - Converts `APIVersionDiscovery` to internal format
- `parse_v2_resource()` / `parse_v2_subresource()` - Parsing helpers
- 10 unit tests for conversion logic
- Integration test (`discovery_run_aggregated`)
- Updated `dynamic_api.rs` example

### Usage

```rust
let client = Client::try_default().await?;
// 2 API calls instead of N+2
let discovery = Discovery::new(client).run_aggregated().await?;
```

### Notes

- Requires Kubernetes 1.26+ (beta) or 1.30+ (stable)
- freshness: "Stale" is currently ignored

doxxx93 commented on kube-client/src/discovery/parse.rs:
```diff
@@ -85,4 +88,239 @@ impl GroupVersionData {
+    res: &APIResourceDiscovery,
+    group: &str,
+    version: &str,
+    api_version: &str,
```
good! thanks!

clux commented on examples/dynamic_api.rs:
```diff
@@ -12,7 +12,8 @@ async fn main() -> anyhow::Result<()> {
 
-    let discovery = Discovery::new(client.clone()).run().await?;
+    // Uses Aggregated Discovery API for fewer API calls (2 instead of N+2)
+    let discovery = Discovery::new(client.clone()).run_aggregated().await?;
```
we should probably swap over the `kubectl` example also

clux commented on kube-client/src/discovery/apigroup.rs:
```diff
@@ -327,7 +361,145 @@ impl ApiGroup {
+                    version: Some("v1beta1".to_string()),
+                    resources: vec![make_v2_resource("jobs", "Job", "Namespaced", vec!["get"])],
+                    freshness: Some("Current".to_string()),
+                },
```
nit: this test might break on us in the future it because it is verifying a property about beta versions. [they removed this version for cronjob](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-25) earlier, so this one is not safe.

on the other hand major versions are meant to [never be removed[(https://kubernetes.io/docs/reference/using-api/deprecation-policy/) (as per `Rule #4a`). So maybe it's better to check the `autoscaling` group which has 2 separate major versions?

clux commented on kube-client/src/discovery/mod.rs:
```diff
@@ -128,6 +128,73 @@ impl Discovery {
+    /// Runs discovery using the Aggregated Discovery API
+    ///
+    /// This method uses the Aggregated Discovery API (available since Kubernetes 1.26, stable in 1.30)
+    /// to fetch all API resources in just 2 requests instead of N+2 requests.
```
Might be worth putting a warning on `run()` above also.

clux commented on kube-client/src/discovery/apigroup.rs:
```diff
@@ -124,6 +125,39 @@ impl ApiGroup {
+            .collect();
+
+        // Preferred version is the first one in the list (they're sorted by preference)
+        let preferred = ag.versions.first().and_then(|v| v.version.clone());
```
nice. good comment. that keeps things simple.

clux reviewed (approved):
This is great. Implementation is clean, mostly private, tests are great, easily readable, docs everywhere doctests. It fits into the existing abstraction with no breaking changes afaikt. Thank you so much for this.

I tried it against a slower cluster (on a different continent) and the general `run` -> `run_aggregated` command went from 28s to 2s so it's a very nice win (particuarly since we provide no caching for the discovery module - maybe something for later).

I only left a few tiny nits and requests in there, but this one is very easy to merge.

doxxx93 commented:
Thanks for the detailed review! All addressed.
- Switched kubectl example to use `run_aggregated()`
- Changed test to use `autoscaling` v1/v2 (stable major versions)
- Added note on `run()` recommending `run_aggregated()`
- Converted `from_v2` to take ownership to avoid clones

clux commented:
Thanks again. I have invited you to a team in the org that gives you slightly more access. No expectations for it, feel free to decline, it's largely symbolic, but you are doing great work and it's there for you, if you'd like the org to show up on your account with some extra access.

doxxx93 commented:
Thank you! I'm honored to be part of the org. Looking forward to contributing more!
